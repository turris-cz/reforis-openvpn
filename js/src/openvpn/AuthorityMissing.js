/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useContext } from "react";
import PropTypes from "prop-types";

import { Button, useAPIPost, AlertContext } from "foris";

import API_URLs from "API";

AuthorityMissing.propTypes = {
    onReload: PropTypes.func.isRequired,
};

export default function AuthorityMissing({ onReload }) {
    const setAlert = useContext(AlertContext);

    const [createResponse, createCA] = useAPIPost(API_URLs.authority);
    useEffect(() => {
        if (createResponse.isSuccess) {
            onReload();
        } else if (createResponse.isError) {
            setAlert(_("Cannot generate certificate authority"));
        }
    }, [createResponse, onReload, setAlert]);

    return (
        <>
            <h3>{_("No certification authority")}</h3>
            <p>{_("Currently there is no OpenVPN certificate authority (CA). A CA is required to generate client certificates to authenticate to the OpenVPN server. To proceed you need to generate it first.")}</p>
            <Button onClick={() => createCA()}>{_("Generate CA")}</Button>
        </>
    );
}
