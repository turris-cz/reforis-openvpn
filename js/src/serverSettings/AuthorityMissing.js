/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { Button, useAPIPost, useAlert, API_STATE } from "foris";

import API_URLs from "API";

AuthorityMissing.propTypes = {
    onReload: PropTypes.func.isRequired,
};

export default function AuthorityMissing({ onReload }) {
    const [setAlert] = useAlert();

    const [createResponse, createCA] = useAPIPost(API_URLs.authority);
    useEffect(() => {
        if (createResponse.state === API_STATE.SUCCESS) {
            onReload();
        } else if (createResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot generate certificate authority"));
        }
    }, [createResponse, onReload, setAlert]);

    return (
        <div className="card p-4">
            <h2>{_("Certificate Authority")}</h2>
            <p>
                {_(
                    "Currently there is no OpenVPN certificate authority (CA). A CA is required to generate client certificates to authenticate to the OpenVPN server. To proceed you need to generate it first."
                )}
            </p>
            <p className="text-center text-muted">
                {_("No certification authority.")}
            </p>
            <div className="text-right">
                <Button
                    onClick={() => createCA()}
                    forisFormSize
                    loading={createResponse.state === API_STATE.SENDING}
                >
                    {_("Generate CA")}
                </Button>
            </div>
        </div>
    );
}
