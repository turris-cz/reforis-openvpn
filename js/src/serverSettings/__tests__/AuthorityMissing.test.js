/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, getByText, fireEvent, wait } from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from 'jest-mock-axios';

import AuthorityMissing from "../AuthorityMissing";

describe("<AuthorityMissing />", () => {
    let componentContainer;
    const handleReload = jest.fn();

    beforeEach(() => {
        const { container } = render(<AuthorityMissing onReload={handleReload} />);
        componentContainer = container;
    });

    it("should match snapshot", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should send request when button is clicked", () => {
        fireEvent.click(getByText(componentContainer, "Generate CA"));
        expect(mockAxios.post).toBeCalledWith("/reforis/openvpn/api/authority", undefined, expect.anything());
    });

    it("should handle error", async () => {
        fireEvent.click(getByText(componentContainer, "Generate CA"));
        mockJSONError();
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith("Cannot generate certificate authority");
        });
    });

    it("should handle success", async () => {
        fireEvent.click(getByText(componentContainer, "Generate CA"));
        mockAxios.mockResponse({});
        await wait(() => {
            expect(handleReload).toHaveBeenCalledTimes(1);
        });
    });
});
