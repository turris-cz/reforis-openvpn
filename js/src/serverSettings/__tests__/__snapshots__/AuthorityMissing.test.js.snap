// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorityMissing /> should match snapshot 1`] = `
<div>
  <div
    class="card p-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      Currently there is no OpenVPN certificate authority (CA). A CA is required to generate client certificates to authenticate to the OpenVPN server. To proceed you need to generate it first.
    </p>
    <p
      class="text-center text-muted"
    >
      No certification authority.
    </p>
    <div
      class="text-right"
    >
      <button
        class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        type="button"
      >
        <span>
          Generate CA
        </span>
      </button>
    </div>
  </div>
</div>
`;
